src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java

```diff
@@ -123,55 +131,114 @@ protected void setState(State state)
+        callback.accept(result, null);
+        state = State.FINISHED;
+    }
+
+    private void fail(Throwable throwable)
+    {
+        Preconditions.checkArgument(state != State.FINISHED && state != State.FAILED);
+        callback.accept(null, throwable);
+        state = State.FAILED;
+    }
+
+    private static PreExecuteContext preExecuteContext(PreLoadContext preLoadContext,
+                                                       AccordStateCache.Instance<TxnId, Command> commandCache,
+                                                       AccordStateCache.Instance<RoutableKey, CommandsForKey> cfkCache)
+    {
+        ImmutableMap<TxnId, Command> ctxCommands = commandCache.getActive(preLoadContext.txnIds());
+        ImmutableMap<RoutableKey, CommandsForKey> ctxCommandsForKey = cfkCache.getActive(toRoutableKeys(preLoadContext.keys()));
+        return new PreExecuteContext()
+        {
+            @Override
+            public ImmutableMap<TxnId, Command> commands()
+            {
+                return ctxCommands;
+            }
+
+            @Override
+            public ImmutableMap<RoutableKey, CommandsForKey> commandsForKey()
+            {
+                return ctxCommandsForKey;
+            }
+
+            @Override
+            public Iterable<TxnId> txnIds()
+            {
+                return preLoadContext.txnIds();
+            }
+
+            @Override
+            public Seekables<?, ?> keys()
+            {
+                return preLoadContext.keys();
+            }
+        };
+    }
+
+    private static <K, V extends ImmutableState> void releaseResources(AccordStateCache.Instance<K, V> cache,
+                                                                       java.util.Map<K, ContextValue<V>> values)
+    {
+         values.forEach((key, value) -> {
+             cache.release(key, value.original(), value.current());
+         });
+    }
+
+    private void releaseResources()
+    {
+        releaseResources(commandStore.commandCache(), postExecuteContext.commands);
+        releaseResources(commandStore.commandsForKeyCache(), postExecuteContext.commandsForKey);
+    }
+
     protected void runInternal()
     {
-        SafeAccordCommandStore safeStore = commandStore.safeStore(context);
         switch (state)
         {
             case INITIALIZED:
                 state = State.LOADING;
             case LOADING:
-                if (!loader.load(context, this))
+                if (!loader.load(this::callback))
                     return;
 
                 state = State.RUNNING;
+                preExecuteContext = preExecuteContext(preLoadContext, commandStore.commandCache(), commandStore.commandsForKeyCache());
+                SafeCommandStore safeStore = commandStore.beginOperation(preExecuteContext);
                 result = apply(safeStore);
+                postExecuteContext = commandStore.completeOperation(safeStore);
```

Comment:
This isn't safe as `apply` has no defensive checks against exceptions, so if an exception is thrown then we never call `commandStore.completeOperation` leaving the `CommandStore` in a broken state (can't start/stop operations as a safe store leaked)

Reply:
looking at the callers in accord-core there are cases that think its fine to throw, which means we need defensive checks in those cases.

Reply:
with this pattern I wonder if `SafeCommandStore` should implement `AutoClosable` so we can get warnings on failure to cleanup; atm this is the only call paths but the API does allow more